components/brave_wallet/browser/asset_discovery_manager.cc

```diff
@@ -48,16 +53,141 @@ AssetDiscoveryManager::GetAssetDiscoverySupportedChains() {
+      continue;
+    }
+
+    if ((*solana_address).IsValid()) {
+      solana_addresses.push_back(*solana_address);
+    }
+  }
+
+  if (solana_addresses.empty()) {
+    CompleteDiscoverAssets(
+        mojom::kSolanaMainnet, std::vector<mojom::BlockchainTokenPtr>(),
+        absl::nullopt, l10n_util::GetStringUTF8(IDS_WALLET_INVALID_PARAMETERS),
+        triggered_by_accounts_added);
+    return;
+  }
+
+  // TODO(nvonpentz): When custom networks are supported, we need to check
+  // that the active network is our own that supports this RPC call.
+
+  const auto barrier_callback =
+      base::BarrierCallback<std::vector<SolanaAddress>>(
+          solana_addresses.size(),
+          base::BindOnce(&AssetDiscoveryManager::MergeDiscoveredSolanaAssets,
+                         weak_ptr_factory_.GetWeakPtr(),
+                         triggered_by_accounts_added));
+  for (const auto& account_address : solana_addresses) {
+    json_rpc_service_->GetSolanaTokenAccountsByOwner(
+        account_address,
+        base::BindOnce(&AssetDiscoveryManager::OnGetSolanaTokenAccountsByOwner,
+                       weak_ptr_factory_.GetWeakPtr(),
+                       std::move(barrier_callback)));
+  }
+}
+
+void AssetDiscoveryManager::OnGetSolanaTokenAccountsByOwner(
+    base::OnceCallback<void(std::vector<SolanaAddress>)> barrier_callback,
+    const std::vector<SolanaAccountInfo>& token_accounts,
+    mojom::SolanaProviderError error,
+    const std::string& error_message) {
+  if (error != mojom::SolanaProviderError::kSuccess || token_accounts.empty()) {
+    std::move(barrier_callback).Run(std::vector<SolanaAddress>());
+    return;
+  }
+
+  // Add each token account to the all_discovered_contract_addresses list
+  std::vector<SolanaAddress> discovered_mint_addresses;
+  for (const auto& token_account : token_accounts) {
+    // Decode Base64
+    const absl::optional<std::vector<uint8_t>> data =
+        base::Base64Decode(token_account.data);
+    if (data.has_value()) {
+      // Decode the address
+      const absl::optional<SolanaAddress> mint_address =
+          DecodeMintAddress(data.value());
+      if (mint_address.has_value()) {
+        // Add the contract address to the list
+        discovered_mint_addresses.push_back(mint_address.value());
+      }
+    }
+  }
+
+  std::move(barrier_callback).Run(discovered_mint_addresses);
+}
+
+void AssetDiscoveryManager::MergeDiscoveredSolanaAssets(
+    bool triggered_by_accounts_added,
+    const std::vector<std::vector<SolanaAddress>>&
+        all_discovered_contract_addresses) {
+  // Create unique flat_set of all discovered contract addresses
+  base::flat_set<std::string> discovered_mint_addresses;
+  for (const auto& discovered_contract_address_list :
+       all_discovered_contract_addresses) {
+    for (const auto& discovered_contract_address :
+         discovered_contract_address_list) {
+      discovered_mint_addresses.insert(discovered_contract_address.ToBase58());
+    }
+  }
+
```

Comment:
Complexity of this is O(n^2). 
Makes sense to collect everything in std::vector and convert inplace to flat_set when binding to OnGetSolanaTokenRegistry callback